File: leetcode/2105-watering-plants-ii/minimumRefill.go
package leetcode

func minimumRefill(plants []int, capacityA int, capacityB int) int {
	var refill = 0
	var (
		indexA = 0
		indexB = len(plants) - 1
		restA  = capacityA
		restB  = capacityB
	)

	for indexA < indexB {
		var plantA = plants[indexA]
		var plantB = plants[indexB]

		if restA >= plantA {
			restA = restA - plantA
		} else {
			refill++

			restA = capacityA - plantA
		}

		if restB >= plantB {
			restB = restB - plantB
		} else {
			refill++

			restB = capacityB - plantB
		}

		indexA++
		indexB--
	}

	if indexA == indexB {
		var rest = max(restA, restB)
		if rest < plants[indexA] {
			refill++
		}
	}

	return refill
}

File: leetcode/2105-watering-plants-ii/minimumRefill.rs
use crate::Solution;

impl Solution {
    pub fn minimum_refill(plants: Vec<i32>, capacity_a: i32, capacity_b: i32) -> i32 {
        let mut refill = 0;
        let mut index_a: usize = 0;
        let mut index_b = plants.len() - 1;
        let mut rest_a = capacity_a;
        let mut rest_b = capacity_b;

        while index_a < index_b {
            let plant_a = plants[index_a];
            let plant_b = plants[index_b];
            if rest_a >= plant_a {
                rest_a = rest_a - plant_a;
            } else {
                refill = refill + 1;
                rest_a = capacity_a - plant_a;
            }

            if rest_b >= plant_b {
                rest_b = rest_b - plant_b
            } else {
                refill = refill + 1;
                rest_b = capacity_b - plant_b;
            }

            index_a = index_a + 1;
            index_b = index_b - 1;
        }

        if index_a == index_b {
          let rest = rest_a.max(rest_b);

          if rest < plants[index_a] {
              refill = refill + 1;
          }
        }

        return refill;
    }
}
